lotus-soup/testkit/node_time.go

```diff
@@ -0,0 +1,120 @@
+package testkit
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/specs-actors/actors/abi"
+	"github.com/raulk/clock"
+	"github.com/testground/sdk-go/sync"
+)
+
+// ClockSyncTopic is the topic where clock synchronization events are exchanged.
+var ClockSyncTopic = sync.NewTopic("clock_sync", &ClockSyncMsg{})
+
+// ClockSyncMsg is a clock synchronization message.
+type ClockSyncMsg struct {
+	ID    string // ${groupname}_${groupseq}
+	Epoch abi.ChainEpoch
+}
+
+// SynchronizeClock synchronizes this instance's clock with the clocks of all
+// miner and client nodes participating in this test scenario.
+//
+// This function returns two channels.
+//
+// The first channel (localAdvance) is where the user requests local advancement
+// to the next epoch. Such requests likely originate from some local event that
+// implies readiness to advance, such as "mining round completed" events from
+// synchronized mining. Upon receiving a local advance signal, the control
+// goroutine forwards the local clock by build.BlockDelaySecs, thus progressing
+// one chain epoch. It then signals this fact to all other nodes participating
+// in the test via the Testground synchronization service.
+//
+// The second channel is where the global clock synchronizer signals every time
+// the entirety of test participants have progressed to the next epoch, thus
+// indicating that all nodes are eady to commence that chain epoch of
+// computation.
+//
+// Essentially this behaviour can be assimilated to a distributed sempaphore.
+//
+// This method will return immediately, and will spawn a background goroutine
+// that performs the clock synchronisation work.
+func (n *LotusNode) SynchronizeClock(ctx context.Context) (localAdvance chan<- abi.ChainEpoch, globalEpoch <-chan abi.ChainEpoch, err error) {
```

Comment:
global vs local: i have trouble grappling with the topology implied by this locality

`the user requests local advancement`: it doesn't seem to me the user (miner) requests advancement, it just says 
i've synced to this epoch, yes, you can interpret it as "i'm done let's move one", but the local/global implies to me some equality that is differentiated between how we order the group, but here we have miners and a central time/chain _control_, the miners don't have any leverage on the control (in more advance use cases even if the miner says i'm done in this round we may still not advance until some other event happens), i don't see the relation that could classify them into local/global with a centralized control system

Reply:
`SynchronizeClock` is another actor (actor-oriented programming). It encapsulates all the control of the clock. No other process in the system (not even the `SynchronizeMiner` actor) is entitled to change the clock (although we don't enforce that prohibition). The idea behind this is that when the local node, when enlisted predicates are satisfied, decides that it's ready to advance the clock, it would send a signal to `localAdvance`.

Right now we ACTUALLY advance the local clock and advertise that fact to the world. In the future, we could use some form of 3PC (three phase commit), where we PREPARE [phase 1] the advancement, nodes opt in or opt out of that round via the sync service, and once all nodes have PREPARED, we actually COMMIT [phase 2] the local clock advancement by manipulating the mock clock, and CERTIFY [phase 3] that we are at the new epoch.

Once all nodes have passed the `CERTIFY` phase, that's when we release the signal on `globalEpoch` to trigger mining and/or other downstream processes.

Reply:
Yes, I think I was having the 3PC model you're describing in my mind. I'm fine if this is not implemented right now.